modules/rostests/apitests/dbghelp/pdb.c

```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
+
+    if (fileinfo == NULL)
+        return FALSE;
+
+    if (hResInfo && dwSize)
+    {
+        UINT uLen;
+
+        HGLOBAL hResData = LoadResource(mod, hResInfo);
+        LPVOID pRes = LockResource(hResData);
+        HLOCAL pResCopy = LocalAlloc(LMEM_FIXED, dwSize);
+        VS_FIXEDFILEINFO *lpFfi;
+
+        CopyMemory(pResCopy, pRes, dwSize);
+        FreeResource(hResData);
+
+        if (VerQueryValueW(pResCopy, L"\\", (LPVOID*)&lpFfi, &uLen))
+        {
+            *fileinfo = *lpFfi;
+            res = TRUE;
+        }
+
+        LocalFree(pResCopy);
+    }
+
+    return res;
+}
+
+static VS_FIXEDFILEINFO dbghelpFileVer;
+static void init_dbghelp_version()
+{
+    LPAPI_VERSION v;
+    WCHAR filenameW[MAX_PATH + 1];
+    HMODULE hDLL;
+    DWORD fileLen;
+    VS_FIXEDFILEINFO fileInfo;
+
+    memset(&dbghelpFileVer, 0, sizeof(dbghelpFileVer));
+
+    /* get internal file version */
+    v = ImagehlpApiVersion();
+    if (v == NULL)
+        return;
+
+    /* get module file version */
+    hDLL = GetModuleHandleW(L"dbghelp.dll");
+    if (hDLL == 0)
+    {
+        ok(FALSE, "Dbghelp.dll is not loaded!\n");
+        return;
+    }
+    if (!get_module_version(hDLL, &fileInfo))
+        memset(&fileInfo, 0, sizeof(fileInfo));
+    dbghelpFileVer = fileInfo;
+
+    /* get full file path */
+    fileLen = GetModuleFileNameW(hDLL, filenameW, MAX_PATH + 1);
+    if (fileLen == 0)
+    {
+        ok(FALSE, "GetModuleFileNameW for dbghelp.dll failed!\n");
+        return;
+    }
+
+    trace("Using %S\n", filenameW);
+    trace("  API-Version: %hu.%hu.%hu (%hu)\n",
+          v->MajorVersion, v->MinorVersion, v->Revision, v->Reserved);
+
+    trace("  Fileversion: %hu.%hu.%hu.%hu\n",
+          HIWORD(fileInfo.dwProductVersionMS),
+          LOWORD(fileInfo.dwProductVersionMS),
+          HIWORD(fileInfo.dwProductVersionLS),
+          LOWORD(fileInfo.dwProductVersionLS));
+}
+
+static
+int g_SymRegisterCallbackW64NotFound = 0;
+
+static
+BOOL WINAPI SymRegisterCallbackW64_Stub(HANDLE hProcess, PSYMBOL_REGISTERED_CALLBACK64 CallbackFunction, ULONG64 UserContext)
+{
+    g_SymRegisterCallbackW64NotFound++;
+    return FALSE;
+}
+
+/* A delay-load failure hook will be called when resolving a delay-load dependency (dll or function) fails */
+FARPROC WINAPI DliFailHook(unsigned dliNotify, PDelayLoadInfo pdli)
+{
+    /* Was the failure a function, and did we get info */
+    if (dliNotify == dliFailGetProc && pdli)
+    {
+        /* Is it our function? */
+        if (pdli->dlp.fImportByName && !strcmp(pdli->dlp.szProcName, "SymRegisterCallbackW64"))
+        {
+            /* Redirect execution to the stub */
+            return (FARPROC)SymRegisterCallbackW64_Stub;
+        }
+    }
+    /* This is not the function you are looking for, continue default behavior (throw exception) */
+    return NULL;
+}
+
+/* Register the failure hook using the magic name '__pfnDliFailureHook2'. */
+PfnDliHook __pfnDliFailureHook2 = DliFailHook;
```

Comment:
What is this good for?

Reply:
Not all dbghelp.dll version have SymRegisterCallbackW64.
This hook gets called if the delayed loaded import for SymRegisterCallBackW64 fails. In this case we can return a stub and prevent an app crash.

Reply:
Using GetProcAddress and bailing out would be much simpler... Why do we delay-load the DLL we are testing anyway ?

Reply:
This looks like a ton of code...